File: weekOfOct23/largestValues.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def largestValues(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """

        if not root:
            return []

        largest = [root.val]
        current = []

        if root.left:
            current.append(root.left)
        if root.right:
            current.append(root.right)

        while current:
            nextLevel = []
            maxVal = current[0].val

            for node in current:
                if node.left:
                    nextLevel.append(node.left)
                if node.right:
                    nextLevel.append(node.right)
                if node.val > maxVal:
                    maxVal = node.val

            largest.append(maxVal)
            current = nextLevel

        return largest

File: weekOfOct23/largestValues.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var largestValues = function(root) {
  if (!root) return []

  const largest = [root.val]
  let current = []

  if (root.left) current.push(root.left)
  if (root.right) current.push(root.right)

  while (current.length) {
      let next = []
      let max = current[0].val

      for(let i = 0; i < current.length; i++) {
          if (current[i].left) next.push(current[i].left)
          if (current[i].right) next.push(current[i].right)
          if (current[i].val > max) max = current[i].val
      }

      largest.push(max)
      current = next
  }

  return largest
};